File: vol2/ch18/LeetCode1162.py
class Solution:
    def maxDistance(self, grid: List[List[int]]) -> int:
        self.dx=[0,0,1,-1]                     #水平方向偏移量
        self.dy=[1,-1,0,0]                      #垂直方向偏移量
        self.qu=deque()
        n=len(grid)
        self.visited=[[False for _ in range(0,n)] for _ in range(0,n)]
        cnt=0
        for i in range(0,n):
            for j in range(0,n):
                if grid[i][j]==1:            #陆地单元
                    self.qu.append([i,j])
                    self.visited[i][j]=True
                    cnt+=1
        if cnt==n*n:return -1     #全部为陆地的情况
        else: return self.bfs(grid)
    def bfs(self,grid):    #多起点分层次广度优先搜索
        n=len(grid)
        step=0
        while self.qu:
            cnt=len(self.qu)
            for i in range(0,cnt):
                [x,y]=self.qu.popleft()
                for di in range(0,4):
                    nx=x+self.dx[di]
                    ny=y+self.dy[di]
                    if nx<0 or nx>=n or ny<0 or ny>=n:
                        continue
                    if self.visited[nx][ny]:continue
                    if grid[nx][ny]==1: continue
                    self.visited[nx][ny]=True
                    self.qu.append([nx,ny])
            step+=1
        return step-1

File: vol2/ch18/LeetCode1162.cpp
class Solution {
    int dx[4]={0,0,1,-1};                        //水平方向偏移量
    int dy[4]={1,-1,0,0};                       //垂直方向偏移量
    queue<pair<int,int>> qu;
    vector<vector<bool>> visited;
public:
    int maxDistance(vector<vector<int>>& grid) {
        int n=grid.size();
        visited=vector<vector<bool>>(n,vector<bool>(n,false));
        int cnt=0;
        for (int i=0;i<n;i++) {
            for (int j=0;j<n;j++) {
                if (grid[i][j]==1) {            //陆地单元
                    qu.push(pair<int,int>(i,j));
                    visited[i][j]=true;
                    cnt++;
                }
            }
        }
        if(cnt==n*n) return -1;     //全部为陆地的情况
        else return bfs(grid);
    }
    int bfs(vector<vector<int>>&grid) {    //多起点分层次广度优先搜索
        int n=grid.size();
        int step=0;
        while(!qu.empty()) {
            int cnt=qu.size();
            for(int i=0;i<cnt;i++) {
                auto e=qu.front();qu.pop();
                int x=e.first,y=e.second;
                for(int di=0;di<4;di++) {
                    int nx=x+dx[di];
                    int ny=y+dy[di];
                    if (nx<0 || nx>=n || ny<0 || ny>=n)
                        continue;
                    if(visited[nx][ny]) continue;		//跳过已访问的单元
                    if(grid[nx][ny]==1) continue;		//跳过陆地单元
                    visited[nx][ny]=true;
                    qu.push(pair<int,int>(nx,ny));
                }
            }
            step++;												//扩展层次增1
        }
        return step-1;
    }
};

